src/wallet/wallet.cpp

```diff
@@ -426,27 +442,289 @@ ZcashdUnifiedSpendingKey CWallet::GenerateNewUnifiedSpendingKey() {
+
+        // Save the metadata for the generated address so that we can re-derive
+        // it in the future.
+        ZcashdUnifiedAddressMetadata addrmeta(ufvkid, diversifierIndex, receiverTypes);
+        if (fFileBacked && !CWalletDB(strWalletFile).WriteUnifiedAddressMetadata(addrmeta)) {
+            throw std::runtime_error(
+                    "CWallet::AdddUnifiedAddress(): Writing unified address metadata failed");
+        }
+
+        if (hasTransparent) {
+            // Regenerate the secret key for the transparent address and add it to
+            // the wallet.
+            auto seed = GetMnemonicSeed().value();
+            auto b44 = libzcash::Bip44AccountChains::ForAccount(seed, BIP44CoinType(), accountId).value();
+            auto key = b44.DeriveExternal(diversifierIndex.ToTransparentChildIndex().value()).value();
+            AddTransparentSecretKey(seed.Fingerprint(), key);
+        }
+
+        return std::make_pair(address.value(), diversifierIndex);
+    } else {
+        return AddressGenerationError::NoSuchAccount;
+    }
+}
+
+// NOTE: There is an important relationship between `LoadUnifiedFullViewingKey`
+// and `LoadUnifiedAddressMetadata`. In all cases, by the end of the
+// wallet-loading process, `mapUfvkAddressMetadata` needs to be fully populated
+// for each unified full viewing key. However, we cannot guarantee in what
+// order UFVKs and unified address and account metadata are loaded from disk.
+// Therefore:
+// * When we load a unified full viewing key from disk, we repopulate
+//   the cache of generated addresses by regenerating the address for each
+//   (diversifierIndex, {receiverType...}) pair in `mapUfvkAddressMetadata`
+//   that we have loaded so far.
+// * When we load a unified address metadata record from disk:
+//   - if we have not yet loaded the unified full viewing key, simply add
+//     an entry `mapUfvkAddressMetadata` so that we can restore the address
+//     once the UFVK has been loaded.
+//   - if we have already loaded the unified full viewing key from disk,
+//     regenerate the address from its metadata and add it to the keystore
+//
+// While this is somewhat complicated, it has the benefit of making each
+// piece of unified full viewing key and address metadata write-once in
+// the wallet database; we never need to update ufvk or address metadata
+// once written.
+bool CWallet::LoadUnifiedFullViewingKey(const libzcash::UnifiedFullViewingKey &key)
+{
+    auto zufvk = ZcashdUnifiedFullViewingKey::FromUnifiedFullViewingKey(Params(), key);
+    auto metadata = mapUfvkAddressMetadata.find(zufvk.GetKeyID());
+    if (metadata != mapUfvkAddressMetadata.end()) {
+        // restore unified addresses that have been previously generated to the
+        // keystore
+        for (const auto &[j, receiverTypes] : metadata->second.GetAllReceivers()) {
+            auto addr = zufvk.Address(j, receiverTypes).value();
```

Comment:
Yeah, looking at how `GetAllReceivers()` is used (assuming this is the only location), I think it would make more sense as:
```suggestion
        for (const auto &[j, addr] : metadata->second.GetAddresses(zufvk)) {
```

Reply:
Your suggestion here won't work, you'd be trying to call `GetAddresses` for addresses whose metadata has been restored but where the addresses themselves have not been added back to the keystore.